plutus-ledger-api/src/Plutus/V1/Ledger/Api.hs

```diff
@@ -0,0 +1,196 @@
+but with the new types and functions underneath. Abstraction lets you do this easily. But we actually
+want to keep the old *implementation*, because things really have to work the same, bug-for-bug. And
+the types have to translate into Plutus Core in exactly the same way, and so on.
+
+So we're going to end up with multiple versions of the types and functions that we expose here, even
+internally. That means we don't lose anything by exposing all the details: we're never going to remove
+anything, we're just going to create new versions.
+-}
+
+-- | Check if a 'Script' is "valid". At the moment this just means "deserialises correctly", which in particular
+-- implies that it is (almost certainly) an encoded script and cannot be interpreted as some other kind of encoded data.
+validateScript :: Script -> Bool
+validateScript = isRight . Flat.unflat @Scripts.Script . fromShort
+
+-- | Convenience constructor for a 'CostModel' from a map giving the parameters.
+mkCostModel :: Map String Integer -> Either EvaluationError CostModel
+-- TODO: this is a bit of a hack, replace it with something more principled
+mkCostModel m = case fromJSON $ toJSON m of
+    Success a -> pure a
+    Error e   -> throwError $ CostModelParameterMismatch e
+
+data VerboseMode = Verbose | Quiet
+
+type LogOutput = [Text.Text]
+
+-- | The cost model parameters are modelled coarsely as a set of named, integer parameters. Conversion to the
+-- "real" model type can fail if those don't line up with the parameters that we expect.
+type CostModelParameters = Map String Integer
+
+-- | Scripts to the ledger are serialised bytestrings.
+type Script = ShortByteString
+
+-- | Errors that can be thrown when evaluating a Plutus script.
+data EvaluationError =
+    CekError (UPLC.CekEvaluationException PLC.DefaultUni PLC.DefaultFun) -- ^ An error from the evaluator itself
+    | DeBruijnError PLC.FreeVariableError -- ^ An error in the pre-evaluation step of converting from de-Bruijn indices
+    | CodecError Flat.DecodeException -- ^ A serialisation error
+    | IncompatibleVersionError (PLC.Version ()) -- ^ An error indicating a version tag that we don't support
+    | CostModelParameterMismatch String -- ^ An error indicating that the cost model parameters didn't match what we expected
+    deriving stock (Show, Eq)
+
+instance Pretty EvaluationError where
+    pretty (CekError e)      = prettyClassicDef e
+    pretty (DeBruijnError e) = pretty e
+    pretty (CodecError e) = viaShow e
+    pretty (IncompatibleVersionError actual) = "This version of the Plutus Core interface does not support the version indicated by the AST:" <+> pretty actual
+    pretty (CostModelParameterMismatch e) = pretty e
+
+-- | Shared helper for the evaluation functions, deserializes the 'Script' , applies it to its arguments, and un-deBruijn-ifies it.
+mkTermToEvaluate :: (MonadError EvaluationError m) => Script -> [Data] -> m (UPLC.Term UPLC.Name PLC.DefaultUni PLC.DefaultFun ())
```

Comment:
How about turning the `mkTermToEvaluate` into a two-step process?

The previous `validateScript` function can be changed to `validateScript :: Script (Bytestring) -> Maybe ScriptDecoded`, signaling failure to decode with `Nothing`.

, then the `mkTermToEvaluate` is changed to take as input the `ScriptDecoded` instead of a ScriptBytestring.

Reply:
This is deliberate: the ledger team don't actually want to see that extra structure and wanted to have deserialization rolled into evaluation.